examples/metricq_drain.py

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019, ZIH, Technische Universitaet Dresden, Federal Republic of Germany
+#
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without modification,
+# are permitted provided that the following conditions are met:
+#
+#     * Redistributions of source code must retain the above copyright notice,
+#       this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above copyright notice,
+#       this list of conditions and the following disclaimer in the documentation
+#       and/or other materials provided with the distribution.
+#     * Neither the name of metricq nor the names of its contributors
+#       may be used to endorse or promote products derived from this software
+#       without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import asyncio
+
+from metricq.subscriber import Subscriber
+from metricq.drain import Drain
+
+
+async def source():
+    print("subscribing")
+    subscriber = Subscriber(
+        "example", "amqp://admin:admin@localhost", metrics=["dummy.source"]
+    )
+    await subscriber.connect()
+
+    print("sleeping...")
+    await asyncio.sleep(10)
+    print("sleeping... over")
+    print("collecting data")
+
+    drain = Drain(
+        "example",
+        "amqp://admin:admin@localhost",
+        metrics=["dummy.source"],
+        queue=subscriber.queue,
+    )
+    await drain.connect()
```

Comment:
Same as above. There should be a convenience method for the `Drain` class as well.

Reply:
Wouldn't it be a good idea to only allow construction of a `Drain` from a subscriber?  That way, one could keep `Subscriber.queue` an implementation detail:

```python
subscriber = Subscriber(...)

... # Do stuff

async with Drain(subscriber, ...) as data:
    print("Received data: {}", data)
```

...where all the magic happens in `Drain.__aenter__` or something.

Reply:
I like that even more!

Reply:
...or one could even turn the dependency around and have:

```python

class Subscriber(...):

    ...

    @asynccontexmanager
    async def drain(self):
       try:
            ...
            yield data
       finally:
            ...


subscriber = Subscriber(...)

... # Do stuff

async with subscriber.drain() as data:
    print("Received data: {}", data)
```

This way there's not even a need to have a dedicated `Drain` class.

Reply:
I'll spitball a bit more.

We could even an interface where we iterate over the data as it arrives:

```python
from asyncio import Queue

class Subscriber:
    ...

    async def drain(self) -> Drain:
        # Unsubscribe from all metrics, set up Drain.
        try:
            drain = ...
        finally:
            await self.stop()
        return drain

class Drain:
    def __init__(self, ...):
        self._data: Queue[tuple[Metric, Timestamp, float]] = Queue()

    def on_data(self, metric, time, value):
        self._data.put_nowait((metric, time, value))

    def __aiter__(self):
        return self

    def __anext__(self):
        # TODO: figure out how end iteration
        return await self._data.get()

subscriber = Subscriber(...)

async for metric, time, value in subscriber.drain():
    print("Asynchronously received data for {metric}: {=time}, {=value}")
```

Reply:
I like the iteration, but your `drain` method won't work like that. The important bit of the `Subscriber` / `Drain` pair is that there is no active connection during the measurement. Thus, in the `Subscriber`, `self.stop()` is called already during the `connect()`, but returning a ready-to-go `Drain` from `Subscriber.drain()` is in order.

For the iteration part, there comes a special message as the last message in the queue that indicates that there won't be any further messages. Upon arrival of this packet, setting a flag should do.